core/clustersmngr/factory.go

```diff
@@ -188,13 +203,61 @@ func (cf *clientsFactory) GetImpersonatedClient(ctx context.Context, user *auth.
 	}
 
 	pool := cf.newClustersPool(cf.scheme)
+	errChan := make(chan error, len(cf.clusters.Get()))
+
+	var wg sync.WaitGroup
+
+	var mutex sync.Mutex
 
 	for _, cluster := range cf.clusters.Get() {
-		if err := pool.Add(ClientConfigWithUser(user), cluster); err != nil {
-			return nil, fmt.Errorf("failed adding cluster client to pool: %w", err)
+		wg.Add(1)
+
+		go func(cluster Cluster, pool ClientsPool, mutex *sync.Mutex, errChan chan error) {
+			defer wg.Done()
+
+			if err := pool.AddSync(ClientConfigWithUser(user), cluster, mutex); err != nil {
+				errChan <- &ClientError{ClusterName: cluster.Name, Err: fmt.Errorf("failed adding cluster client to pool: %w", err)}
+			}
+		}(cluster, pool, &mutex, errChan)
+	}
+
+	wg.Wait()
+	close(errChan)
```

Comment:
You close `errChan` here before looping over it on line 229.
It means that you will already get `nil` out of `errChan`.

I'm guessing (correct me if I'm wrong) that this `errChan` would be needed to close via `defer`?

Reply:
Hi @chanwit 

Because the channel is buffered it means we can place messages onto the channel and then close it, and still receive the messages later when we `range` over the channel. See here for an example: https://gobyexample.com/range-over-channels. Here is a playground example in action: https://go.dev/play/p/_tN7jc6tTz2

Reply:
Thank you for sharing the playground example. It makes sense to me now.